### PR TITLE
Optimize topology spreading filter

### LINKED ISSUE #84936: Improve performance of EvenPodSpread (priority and predicate)
In the light of a new performance test that compares EvenPodSpread with SelectorSpread (#84606), it seems that EvenPodSpread can be faster.

Let's investigate if we can achieve parity by carefully building the data structures.

/sig scheduling

### PR DESCRIPTION
**Pull request by alculquicondor**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

There are 2 optimizations in this PR:
- Use a map of pointers and atomic changes to keep the counter per node
- Set an initial capacity for the map

Benchmark results:

```
master

BenchmarkFilter/1000nodes/single-constraint-zone-56         	    1108	   1036523 ns/op
BenchmarkFilter/1000nodes/single-constraint-node-56         	     471	   2518077 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node-56      	     367	   3729778 ns/op

+ atomic

BenchmarkFilter/1000nodes/single-constraint-zone-56         	    1377	    999848 ns/op
BenchmarkFilter/1000nodes/single-constraint-node-56         	     662	   1698138 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node-56      	     508	   2269242 ns/op

+ map capacity

BenchmarkFilter/1000nodes/single-constraint-zone-56         	    1317	   1059039 ns/op
BenchmarkFilter/1000nodes/single-constraint-node-56         	     939	   1269569 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node-56      	     660	   1841336 ns/op
```

Note that cloning the state becomes more expensive:

```
master

BenchmarkFilter/1000nodes/single-constraint-zone/Clone-56   	  269688	      4880 ns/op
BenchmarkFilter/1000nodes/single-constraint-node/Clone-56   	    4214	    295056 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node/Clone-56         	    3637	    305096 ns/op

PR

BenchmarkFilter/1000nodes/single-constraint-zone/Clone-56   	  254336	      5195 ns/op
BenchmarkFilter/1000nodes/single-constraint-node/Clone-56   	    3175	    380806 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node/Clone-56         	    2961	    369687 ns/op
```

But since cloning only happens when doing preemption, and the difference is small, it's worth the change.

**Which issue(s) this PR fixes**:

Fixes #84936

**Special notes for your reviewer**:

Note that special-casing hostname topology (as done in #89487) doesn't work for Filter. We need the global minimum for each topology, thus we need to calculate all counters during PreFilter.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by alculquicondor** [comment]
/assign @Huang-Wei 
cc @ahg-g

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 1]
```diff
@@ -219,52 +220,47 @@ func (pl *PodTopologySpread) calPreFilterState(pod *v1.Pod) (*preFilterState, er
...
-		lock.Lock()
-		s.TpPairToMatchNum[pair] += num
-		lock.Unlock()
+		TpPairToMatchNum:     make(map[topologyPair]*int32),
 	}
-
-	processNode := func(i int) {
-		nodeInfo := allNodes[i]
-		node := nodeInfo.Node()
+	for _, n := range allNodes {
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go]: I recalled I did the same attempt but seemed the perf didn't improve much. I will run the benchmark later.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 2]
```diff
@@ -219,52 +220,47 @@ func (pl *PodTopologySpread) calPreFilterState(pod *v1.Pod) (*preFilterState, er
...
+	}
+
+	processNode := func(i int) {
+		nodeInfo := allNodes[i]
+		node := nodeInfo.Node()
 
 		// Ensure current node's labels contains all topologyKeys in 'Constraints'.
 		if !nodeLabelsMatchSpreadConstraints(node.Labels, constraints) {
 			return
 		}
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go]: Based on current logic, this is no longer needed - non-fit node will "continue" `len(constraints)` times and does no-op when hitting:

```go
			tpCount := s.TpPairToMatchNum[pair]
			if tpCount == nil {
				continue
			}
```

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 3]
```diff
@@ -61,14 +61,15 @@ func (s *preFilterState) Clone() framework.StateData {
...
+		TpPairToMatchNum:     make(map[topologyPair]*int32, len(s.TpPairToMatchNum)),
 	}
 	for tpKey, paths := range s.TpKeyToCriticalPaths {
 		copy.TpKeyToCriticalPaths[tpKey] = &criticalPaths{paths[0], paths[1]}
 	}
 	for tpPair, matchNum := range s.TpPairToMatchNum {
 		copyPair := topologyPair{key: tpPair.key, value: tpPair.value}
-		copy.TpPairToMatchNum[copyPair] = matchNum
+		copyCount := *matchNum
+		copy.TpPairToMatchNum[copyPair] = &copyCount
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go]: I believe we should also consider the performance of Clone() as Clone() is used a lot during preemption. Could you benchmark it? (we can tolerate slight perf downgrade, but not much)

I asked this as instantiating a pointer is expensive than we think - we used to hit some regression when submitting performance PRs.

**Review by Huang-Wei — COMMENTED** [review]
Thanks Aldo. Some comments below. And I'm doing some local tests on it.

**Comment by Huang-Wei** [comment]
> And I'm doing some local tests on it.

I did the benchmark test and got similar result. That's cool!

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go]: Yes, I forgot to remove this before

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go [hunk 4]
```diff
@@ -1108,12 +1110,19 @@ func BenchmarkTestCalPreFilterState(b *testing.B) {
...
 			for i := 0; i < b.N; i++ {
-				s := pl.PreFilter(context.Background(), framework.NewCycleState(), tt.pod)
+				state := framework.NewCycleState()
+				s := pl.PreFilter(ctx, state, tt.pod)
 				if !s.IsSuccess() {
 					b.Fatal(s.AsError())
 				}
+				filterNode := func(i int) {
+					n, _ := pl.sharedLister.NodeInfos().Get(allNodes[i].Name)
+					_ = pl.Filter(ctx, state, tt.pod, n)
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go]: I thought my IDE would complain, but actually it only does that for functions that return errors.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go (hunk 3 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go]: I got this:

```
master

BenchmarkFilter/1000nodes/single-constraint-zone/Clone-56   	  269688	      4880 ns/op
BenchmarkFilter/1000nodes/single-constraint-node/Clone-56   	    4214	    295056 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node/Clone-56         	    3637	    305096 ns/op

PR

BenchmarkFilter/1000nodes/single-constraint-zone/Clone-56   	  254336	      5195 ns/op
BenchmarkFilter/1000nodes/single-constraint-node/Clone-56   	    3175	    380806 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node/Clone-56         	    2961	    369687 ns/op
```

Which translates to 23% slower for 1000 nodes. Given that we are still talking about less than 1ms, I'm ok with this. If you agree, I'll include this numbers in the PR description.

Comment (Huang-Wei) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go]: I'm fine with it - leave a comment in the code is also fine.

**Comment by Huang-Wei** [comment]
@alculquicondor thanks for this PR. Please squash to 3 commits - benchmark, use *int64 pointer, and initialize map capacity.

**Comment by alculquicondor** [comment]
squashed

**Pull request merged**